### ISSUE TITLE
Backing up WhatsApp data through the multi-device web client

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8877]
Backing up WhatsApp data through the multi-device web client<br>
<br>
https://ift.tt/povKLR6<br>
<br>
Francesco Mazzoli &lt;f@mazzo.li&gt;<br>
<br>
<div>
<div>
<div>
<div>
<div>
<h1>
<em><span>2021-10-26</span></em> Backing up WhatsApp data through the multi-device web client</h1>
<p><strong>Abstract:</strong> Systematically backing up WhatsApp messages and media in an open format is impractical. However the <a href="https://faq.whatsapp.com/web/download-and-installation/how-to-join-or-leave-the-multi-device-beta/?lang=en">multi-device beta</a> makes the WhatsApp web client an almost first-class WhatsApp client. I describe how the web client stores messages and media, and a proof of concept program, <a href="https://github.com/bitonic/wadump"><code>wadump</code></a>, which can extract and display them.</p>
<hr>
<p>After I wrote the tool and the post, I realized that somebody had <a href="https://github.com/sigalor/whatsapp-web-reveng">already reverse engineered the WhatsApp web client protocol</a>. Some of the information in this post (especially regarding media encryption) is explained more completely in that project.</p>
<hr>
<p>As you might have noticed from some of my other <a href="https://mazzo.li/posts/gdpr-success-story.html">blog</a> <a href="https://mazzo.li/posts/hetzner-zfs.html">posts</a>, I care about controlling my data. This is not out of privacy concerns, but because I want to be sure that I will never lose access to it. This fear is not abstract: I did lose access to a lot of my data a long time ago, and never got over it &#128579;.</p>
<div>
<p>These days I&rsquo;m quite happy with my data situation, with one big exception: WhatsApp. For better of worse a large portion of my communication goes through it, and I would hate to lose some of my WhatsApp text and audio messages. I do back them up to Google Drive using the built-in backup functionality, but that data depends on me having access to my Google account, and is also not usable without the official WhatsApp app.</p>
<p>In other words, I think there&rsquo;s a chance that I won&rsquo;t have access to my &ldquo;official&rdquo; WhatsApp data in a few decades.</p>
<p>On Android other users have decrypted the <code>.crypt12</code> files that WhatsApp uses, but I never bothered to root my phone, which is a requirement to get the decryption keys, and I&rsquo;m generally not very familiar with Android as a platform.</p>
<p>This is why I was very excited to learn that <a href="https://faq.whatsapp.com/general/download-and-installation/about-multi-device-beta/?lang=en">WhatsApp is rolling out a &ldquo;real&rdquo; desktop application</a>, which does not need the phone to operate. Since the desktop application can send and receive messages independently, it must also store them independently. And since the desktop application is just a web application, it should be relatively easy to programmatically extract the messages for backup purposes.</p>
</div>
<div>
<p>This article describes my finding on how the WhatsApp web application stores messages and media. It also describes a program, <a href="https://github.com/bitonic/wadump"><code>wadump</code></a>, which automatically dumps all the WhatsApp data present in the web client, together with a rudimentary viewer for said data:</p>
<div>
<p><video> Your browser does not support the video tag.</video></p>
</div>
<p>If you&rsquo;re in a hurry you can just head directly to <a href="https://github.com/bitonic/wadump">the GitHub repo</a> which contains instructions on how to run it.</p>
</div>
<h2>Disclaimer &amp; Limitations <a href="https://mazzo.li/posts/whatsapp-backup.html#disclaimer-limitations">#</a>
</h2>
<p>I am not affiliated with Facebook or WhatsApp, and this investigation was done purely to preserve my personal data better. I don&rsquo;t know if backing up your data this way breaches WhatsApp&rsquo;s terms of service. Use at your own risk!</p>
<p>This information is useful at the time of writing, but it could very well not be in the near future if the web client changes the way it stores its data.</p>
<p>Finally, <code>wadump</code> currently <a href="https://mazzo.li/posts/whatsapp-backup.html#media-troubles">cannot reliably extract all media</a>.</p>
<h2>How multi-device WhatsApp works <a href="https://mazzo.li/posts/whatsapp-backup.html#how-multi-device-whatsapp-works">#</a>
</h2>
<p>This is explained well in the <a href="https://engineering.fb.com/2021/07/14/security/whatsapp-multi-device/">Facebook Engineering blog post about the multi-device beta</a>, but the key concept is simple.</p>
<p>Before the multi-device web app, the web client relied on the phone to send and receive messages, deferring all E2EE with our contacts to the phone:</p>
<p>The multi-device beta changes this, so that many devices (up to 4) can send and receive messages directly. If we have more than one device, say a laptop and a phone, our contacts will encrypt their messages once per device, and similarly when we send a message from one of our devices we will route it both to the recipient and to our other devices:</p>
<div>
<p>While the changes in how communication between clients are interesting, in this blog post we&rsquo;re interested in how the web client stores local messages, rather.</p>
<p>When the web client first connects, the phone (which remains a &ldquo;privileged&rdquo; client in terms of key and device management) sends an initial bundle of messages to the web client, together with other information such as our contacts. After this initial setup is done, the web client is operational, and it will receive and store all new messages that get sent to our number.</p>
</div>
<h2>How data is stored by the WhatsApp client <a href="https://mazzo.li/posts/whatsapp-backup.html#how-data-is-stored-by-the-whatsapp-client">#</a>
</h2>
<p>The data stored by the web client can be divided in three parts:</p>
<ul>
<li>Metadata stored as cleartext in IndexedDB;</li>
<li>Encrypted text chat content, also stored in IndexedDB;</li>
<li>Encrypted media files, stored in WhatsApp&rsquo;s CDN, and cached locally.</li>
</ul>
<p><a href="https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API">IndexedDB</a> is a persistent storage facility for web applications &mdash; think of an in-browser key-value store. Web applications can create databases and tables (which IndexedDB calls &ldquo;object stores&rdquo;), and store JavaScript objects indexed by some key in them.</p>
<p>The cleartext metadata includes a lot of information, mostly stored in the <code>model-storage</code> database, including:</p>
<ul>
<li>Our contact list, including names and phone numbers, in object store <code>contact</code>;</li>
<li>The list of groups we&rsquo;re in, in object store <code>group-metadata</code>;</li>
<li>Information about the messages we&rsquo;ve sent and received, including senders and recipients, their time, and in which chat they appeared, in object stores <code>message</code> and <code>chat</code>.</li>
</ul>
<p>You can check this data out for yourself in the &ldquo;IndexedDB&rdquo; section of the &ldquo;Application&rdquo; tab of the Chrome dev tools.</p>
<p>The cleartext data can be saved by just <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L384">dumping the relevant object stores</a>. Retrieving the actual contents of the messages, wether text or media, is trickier.</p>
<h2>Decrypting text chats <a href="https://mazzo.li/posts/whatsapp-backup.html#decrypting-text-chats">#</a>
</h2>
<p>Messages are stored as JSON objects in the <code>message</code> object store. For text messages, the content of the message is encrypted using AES-CBC, as implemented by <a href="https://developer.mozilla.org/en-US/docs/Web/API/AesCbcParams">the <code>SubtleCrypto</code> API</a>:</p>
<div>
<pre><code><span>{</span>
<span> <span>id</span><span>:</span> <span>...,</span></span>
<span> <span>t</span><span>:</span> <span>1517917600</span><span>,</span> <span>// seconds epoch</span></span>
<span> <span>from</span><span>:</span> <span>...,</span></span>
<span> <span>to</span><span>:</span> <span>...,</span></span>
<span> <span>msgRowOpaqueData</span><span>:</span> { </span>
<span> <span>// An identifier telling us which key to use to decrypt this message</span></span>
<span> <span>_keyId</span><span>:</span> <span>1</span><span>,</span></span>
<span> <span>// The initialization vector to be used by AES-CBC decryption, see</span></span>
<span> <span>// &lt;https://developer.mozilla.org/en-US/docs/Web/API/AesCbcParams&gt;</span></span>
<span> <span>iv</span><span>:</span> <span>ArrayBuffer</span>(<span>16</span>)<span>,</span></span>
<span> <span>// The actual encrypted data</span></span>
<span> <span>_data</span><span>:</span> <span>ArrayBuffer</span>(<span>80</span>)<span>,</span> </span>
<span> }<span>,</span></span>
<span> <span>...</span> omitted fields <span>...</span></span>
<span>}</span></code></pre>
</div>
<p>AES-CBC is a symmetric encryption scheme &mdash; the same key is used to encrypt and decrypt messages. But where does this key come from?</p>
<h3>Storing, retrieving, and deriving keys <a href="https://mazzo.li/posts/whatsapp-backup.html#storing-retrieving-and-deriving-keys">#</a>
</h3>
<div>
<p>To operate, the web client stores some cryptographic keys in IndexedDB. These keys are needed both to perform E2EE when sending and receiving messages, but also to encrypt the messages we store locally. Presumably this is done so that third-parties cannot read messages even if they have access to our machine &mdash; if the application isn&rsquo;t currently running anyway.</p>
<p>The keys are stored through the <a href="https://developer.mozilla.org/en-US/docs/Web/API/CryptoKey"><code>CryptoKey</code> API</a>, which lets us use the keys in JavaScript without being able to leak their contents, inadvertently or otherwise.</p>
</div>
<p>The keys that we are interested in are in <code>keys</code> object store in the <code>wawc_db_enc</code> IndexedDB database. Each object in this object store looks like this:</p>
<div>
<pre><code><span>{</span>
<span> <span>id</span><span>:</span> number<span>,</span> <span>// the key id, as referenced from the messages</span></span>
<span> <span>key</span><span>:</span> <span>CryptoKey</span><span>,</span></span>
<span> <span>_expiration</span><span>:</span> number<span>,</span> <span>// milliseconds epoch</span></span>
<span>}</span></code></pre>
</div>
<p>In my case I have only one key, with <code>id</code> 1.</p>
<h3>Deriving the right key <a href="https://mazzo.li/posts/whatsapp-backup.html#deriving-the-right-key">#</a>
</h3>
<p>The keys in IndexedDB are not used directly to encrypt and decrypt messages. Rather they are be used to <em>derive</em> an AES-CBC key for message encryption, using the <a href="https://en.wikipedia.org/wiki/HKDF">HKDF</a> key derivation function. Key derivation functions are used to turn some secret (maybe a passphrase or, like in our case, another key) into a key suitable for the task at hand.</p>
<p>HKDF in particular lets us integrate other bits of information (the &ldquo;salt&rdquo; and the &ldquo;info&rdquo;), and allows us to specify the length of the output key.</p>
<p>WhatsApp web uses <a href="https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey">the <code>deriveKey</code> function from <code>SubtleCrypto</code></a> to turn our &ldquo;master&rdquo; key into an AES-CBC key suitable for local symmetric encryption of our text messages. It looks like this:</p>
<div>
<pre><code><span><span>// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey</span></span>
<span>crypto<span>.</span><span>subtle</span><span>.</span><span>deriveKey</span>(</span>
<span> <span>// The algorithm used to derive the key (HKDF with SHA256 as hash</span></span>
<span> <span>// function), together with its parameters -- the salt and the info.</span></span>
<span> <span>// We'll see below where these come from.</span></span>
<span> {</span>
<span> <span>name</span><span>:</span> <span>"HKDF"</span><span>,</span></span>
<span> <span>hash</span><span>:</span> <span>"SHA-256"</span><span>,</span></span>
<span> salt<span>,</span></span>
<span> info<span>,</span></span>
<span> }</span>
<span> <span>// The key stored in `wawc_db_enc`</span></span>
<span> key<span>,</span></span>
<span> <span>// The specification of the derived key</span></span>
<span> { <span>name</span><span>:</span> <span>"AES-CBC"</span><span>,</span> <span>length</span><span>:</span> <span>128</span> }<span>,</span></span>
<span> <span>// Whether the key can be extracted (e.g. whether we can dump its contents</span></span>
<span> <span>// with `crypto.subtle.exportKey`)</span></span>
<span> <span>false</span><span>,</span></span>
<span> <span>// What we'll use the key for -- in this case we'll perform symmetric</span></span>
<span> <span>// encryption with it, to store and retrieve our text messages</span></span>
<span> [ <span>"encrypt"</span><span>,</span> <span>"decrypt"</span> ]<span>,</span></span>
<span>)<span>;</span></span></code></pre>
</div>
<p>Given the code above, all we need to generate the appropriate key is the <code>salt</code> and the <code>info</code> parameters to the HKDF function. The <code>salt</code> is non-secret data which <a href="https://datatracker.ietf.org/doc/html/rfc5869#section-3.1">gives HKDF some desirable properties</a>, while the <code>info</code> parameter <a href="https://datatracker.ietf.org/doc/html/rfc5869#section-3.2">is used to &ldquo;bind&rdquo; the key for a certain application</a>.</p>
<p>Typically the <code>salt</code> is generated at random, while the <code>info</code> contains some info on where the key is used, such as protocol number, user identity, and so on.</p>
<h3>Finding the <code>info</code> and <code>salt</code> <a href="https://mazzo.li/posts/whatsapp-backup.html#finding-the-info-and-salt">#</a>
</h3>
<div>
<p>The <code>info</code> is readily available: it is stored in the browser <a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage"><code>localStorage</code></a>, with key <code>WebEncKeySalt</code>, base64 encoded.</p>
<p>Retrieving the <code>salt</code> requires more work. It is sent in a welcome message on the WebSockets connection that WhatsApp web client uses to communicate with the server.</p>
<p>We could read and interpret the WebSockets message at startup, but that would require writing a parser for WhatsApp&rsquo;s protocol, and making sure to run the code as soon as the web client starts. However, there is a simpler way to access the information we need to decrypt the messages.</p>
</div>
<h3>Retrieving the derived key by monkey patching <a href="https://mazzo.li/posts/whatsapp-backup.html#retrieving-the-derived-key-by-monkey-patching">#</a>
</h3>
<p>WhatsApp web uses <code>crypto.subtle.decrypt</code> to decrypt messages. Luckily, <code>crypto.subtle.decrypt</code> is not a read-only property, and we can <a href="https://en.wikipedia.org/wiki/Monkey_patch">replace it</a> with a function which waits until it is called with arguments which are capable of decoding messages.</p>
<p><a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L342">It looks like this</a>:</p>
<ul>
<li>We read a message from the <code>message</code> object store;</li>
<li>We replace <code>crypto.subtle.decrypt</code> with a function which tries to decode the message above with the arguments passed to decrypt;</li>
<li>Once we&rsquo;ve found a key that works, we store it in a global variable, restore the original function, and start decoding the messages in bulk.</li>
</ul>
<p>This relies on the same key working for all messages, which is the case for me.</p>
<p>Using this method we avoid having to derive the key ourselves, at the cost of requiring some user interaction to trigger message decryption in the web client.</p>
<h3>Decoding the content of the messages <a href="https://mazzo.li/posts/whatsapp-backup.html#decoding-the-content-of-the-messages">#</a>
</h3>
<p>Once <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L275"><code>msgRowOpaqueData._data</code> is decrypted</a>, the resulting cleartext is a <a href="https://developers.google.com/protocol-buffers">Protocol Buffer</a> message with this schema:</p>
<pre><code>message WhatsAppMessage {
  string body = 1;
  string caption = 2;
  string loc = 4;
  double lng = 5;
  double lat = 7;
  int32 paymentAmount1000 = 8;
  string paymentNoteMsgBody = 9;
  string canonicalUrl = 10;
  string matchedText = 11;
  string title = 12;
  string description = 13;
  bytes futureproofBuffer = 14;
}

message FullWhatsAppMessage {
  WhatsAppMessage currentMsg = 1;
  WhatsAppMessage quotedMsg = 1;
}</code></pre>
<p>The <code>quotedMsg</code> contains the message we&rsquo;re replying to, if any. To decode the protobuf messages without depending on external code, <code>wadump</code> contains <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L81">a minimal protobuf decoder</a>.</p>
<p>Now that we&rsquo;ve covered text chats, let&rsquo;s talk about media.</p>
<p>Media (audio messages, images, videos, documents) is predictably not stored inline in the protobuf message above. Rather, it is encrypted and stored in WhatsApp&rsquo;s CDNs, and retrieved and decrypted when needed. A media message looks like this in the <code>message</code> object store:</p>
<div>
<pre><code><span>{</span>
<span> <span>id</span><span>:</span> <span>...,</span></span>
<span> <span>t</span><span>:</span> <span>1627663880</span><span>,</span> <span>// seconds epoch</span></span>
<span> <span>from</span><span>:</span> <span>...,</span></span>
<span> <span>to</span><span>:</span> <span>...,</span></span>
<span> <span>// Might be "video", "audio", etc.</span></span>
<span> <span>type</span><span>:</span> <span>"image"</span><span>,</span></span>
<span> <span>mimetype</span><span>:</span> <span>"image/jpeg"</span><span>,</span></span>
<span> <span>// base64 encoded SHA256 hash of the decrypted message</span></span>
<span> <span>filehash</span><span>:</span> <span>"dzBnGb+PQVMhdlkuoUkXB+3xMWEJsCadLRJ9JVOdTeQ="</span><span>,</span></span>
<span> <span>// base64 encoded key used to decrypt the message</span></span>
<span> <span>mediaKey</span><span>:</span> <span>"e4Q1drkxa4ST0+Ac9Ug86W0etx3WbDaZ0IGRs/3ZVI4="</span><span>,</span></span>
<span> <span>// CDN path to the encrypted message</span></span>
<span> <span>directPath</span><span>:</span> <span>"/v/t62.7118-24/35487369_1391679344547224_9081397384390117319_n.enc?ccb=11-4&amp;oh=a8323bff0917bdc1827929420adec846&amp;oe=61287211&amp;_nc_hot=1627663880"</span><span>,</span></span>
<span> <span>// Image metadata...</span></span>
<span> <span>size</span><span>:</span> <span>165863</span><span>,</span></span>
<span> <span>height</span><span>:</span> <span>1281</span><span>,</span></span>
<span> <span>width</span><span>:</span> <span>1600</span><span>,</span></span>
<span> <span>...</span> omitted fields <span>...</span></span>
<span>}</span></code></pre>
</div>
<p>The main CDN domain for WhatsApp seems to be <code>https://mmg.whatsapp.net</code>. Appending the <code>directPath</code> to it will give us the URL to hit to download the encrypted file. Once that is done, we need to decrypt it.</p>
<p>AES-CBC is used to also decrypt media files. Like with messages, HKDF with SHA256 as hash function is used to derive the key. This time the parameters to HKDF are:</p>
<ul>
<li>The key is the 32 bytes of data encoded in base64 in the <code>mediaKey</code> field of the message object.</li>
<li>The <code>info</code> is derived from the <code>type</code> of the message. For <code>"image"</code> it is <code>"WhatsApp Image Keys"</code>, for <code>"video"</code> it is <code>"WhatsApp Video Keys"</code>, <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L155">and so on</a>.</li>
<li>The <code>salt</code> is not used, in which case HKDF just uses <a href="https://datatracker.ietf.org/doc/html/rfc5869#section-2.2">a dummy salt filled with 0s</a>.</li>
<li>The output key length is 112 bytes.</li>
</ul>
<p>In this case HKDF is <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L205">performed directly on the raw data from <code>mediaKey</code></a>, without going through <code>crypto.subtle.importKey</code> and <code>crypto.subtle.deriveKey</code>.</p>
<p>Once the 112 bytes long key is generated, <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L215">it is chopped up</a> to get the initialization vector and key for AES-CBC decryption:</p>
<div>
<pre><code><span><span>const</span> mediaKeys <span>=</span> {</span>
<span> <span>iv</span><span>:</span> key<span>.</span><span>slice</span>(<span>0</span><span>,</span> <span>16</span>)<span>,</span></span>
<span> <span>encKey</span><span>:</span> key<span>.</span><span>slice</span>(<span>16</span><span>,</span> <span>48</span>)<span>,</span></span>
<span> <span>// Other keys, unused for our purposes</span></span>
<span> <span>macKey</span><span>:</span> key<span>.</span><span>slice</span>(<span>48</span><span>,</span> <span>80</span>)<span>,</span></span>
<span> <span>refKey</span><span>:</span> key<span>.</span><span>slice</span>(<span>80</span><span>,</span> <span>112</span>)</span>
<span>}</span></code></pre>
</div>
<p>Note how in this case, unlike with text messages, we only need cleartext data to decrypt media files.</p>
<p>Once we have the data above, we can go ahead and use it to <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L223">decrypt the encrypted bytes</a>:</p>
<div>
<pre><code><span><span>const</span> key <span>=</span> <span>await</span> crypto<span>.</span><span>subtle</span><span>.</span><span>importKey</span>(<span>"raw"</span><span>,</span> mediaKeys<span>.</span><span>encKey</span><span>,</span> <span>"AES-CBC"</span><span>,</span> <span>false</span><span>,</span> [<span>"decrypt"</span>])<span>;</span></span>
<span><span>// the last 10 bytes contain the mac</span></span>
<span>bytes <span>=</span> bytes<span>.</span><span>slice</span>(<span>0</span><span>,</span> <span>-</span><span>10</span>)<span>;</span></span>
<span><span>const</span> cleartext <span>=</span> <span>await</span> crypto<span>.</span><span>subtle</span><span>.</span><span>decrypt</span>({ <span>name</span><span>:</span> <span>"AES-CBC"</span><span>,</span> <span>iv</span><span>:</span> mediaKeys<span>.</span><span>iv</span> }<span>,</span> key<span>,</span> bytes)<span>;</span></span></code></pre>
</div>
<p>The web client caches decrypted media files in the <code>"lru-media-array-buffer-cache"</code> cache, through the <a href="https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage"><code>CacheStorage</code> API</a>.</p>
<p>When looking for media files, <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L244"><code>wadump</code> looks there first</a>, and stores downloaded files in the cache after download, as the web client would. This is to avoid sending off thousands of CDN requests each time we want to dump the media messages.</p>
<p>While the method above works for recent media files, it seems that the WhatsApp CDN does not store media forever. In those cases requesting the file results in a 410 or 404. When that happens the web client re-requests the media file from the phone and gets a new link from the WebSocket channel.</p>
<p>I have not studied in detail the protocol that is used on the WebSocket channel, and I have not experimented with using it independently from the web client. Therefore <code>wadump</code> can currently only retrieve relatively recent or cached media messages.</p>
<p>Another mystery is how the encrypted <code>msgRowOpaqueData</code> is used for for media messages, since it is still present. It might be used to request the missing files above, but again, I have not figured that out yet.</p>
<h2>Wrapping up the dump, and browsing it <a href="https://mazzo.li/posts/whatsapp-backup.html#wrapping-up-the-dump-and-browsing-it">#</a>
</h2>
<p>Once <code>wadump</code> has collected all the information that it can find, <a href="https://github.com/bitonic/wadump/blob/3cf6b7897e0dd1c07ee275ecc4ca4884463b5dd1/dump-messages.js#L116">it stores everything in a tarball</a>, which it then offers for download.</p>
<p>The tarball contains the following files:</p>
<pre><code>message.json
chat.json
contact.json
group-metadata.json
media/&lt;filehash&gt;</code></pre>
<p>The <code>chat.json</code>/<code>contact.json</code>/<code>group-metadata.json</code> files simply contain a dump of the respective object stores. <code>message.json</code> contains a dump of the <code>message</code> object store, but with the <code>msgRowOpaqueData</code> field replaced by a <code>msgRow</code> field with the decrypted text message.</p>
<p>The <code>media</code> folder contains all the media files we could retrieve, named with their <code>filehash</code>.</p>
<p><code>wadump</code> also contains a small utility to easily browse the contents of the tarball archive &mdash; head to <a href="https://github.com/bitonic/wadump/blob/master/README.md">the README</a> for more information.</p>
<h2>Closing thoughts &amp; future work <a href="https://mazzo.li/posts/whatsapp-backup.html#closing-thoughts">#</a>
</h2>
<div>
<p>It was fun to figure out how the WhatsApp web client works, but currently <code>wadump</code> isn&rsquo;t a tool I would rely on. With enough interest a browser extension could be created to keep the WhatsApp web client data in sync with the local file system, possibly using the new <a href="https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API">File System Access API</a>. This would also work around the limitations in media downloads by downloading new media files soon after they appear in a message.</p>
<p>Obviously writing software targeting a proprietary protocol will always be an uphill struggle, but projects like <a href="https://youtube-dl.org/">youtube-dl</a> prove that with effort it is possible to maintain useful tools of this kind.</p>
</div>
<h2>Acknowledgements <a href="https://mazzo.li/posts/whatsapp-backup.html#acknowledgements">#</a>
</h2>
<p>Thanks to Alexandru Scvor&#355;ov and Alex Appetiti for reading drafts of this article.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Francesco Mazzoli<br>
<br>
October 8, 2022 at 05:06PM
